core/storage/mpool/mpool.cpp

```diff
@@ -20,47 +26,562 @@ namespace fc::storage::mpool {
     std::vector<SignedMessage> messages;
     for (auto &[addr, pending] : by_from) {
-      for (auto &[nonce, message] : pending.by_nonce) {
+      for (auto &[nonce, message] : pending) {
         messages.push_back(message);
       }
     }
     return messages;
   }
 
-  outcome::result<Nonce> Mpool::nonce(const Address &from) const {
+  outcome::result<std::pair<std::vector<TipsetCPtr>, std::vector<TipsetCPtr>>>
+  findPath(TsLoadPtr ts_load, TipsetCPtr from, TipsetCPtr to, size_t depth) {
+    std::vector<TipsetCPtr> revert, apply;
+    while (true) {
+      if (from->key == to->key) {
+        return std::make_pair(std::move(revert), std::move(apply));
+      }
+      if (apply.size() > depth || revert.size() > depth) {
+        return ERROR_TEXT("findPath: too deep");
+      }
+      if (from->height() < to->height()) {
+        apply.push_back(to);
+        OUTCOME_TRYA(to, ts_load->load(to->getParents()));
+      } else {
+        revert.push_back(from);
+        OUTCOME_TRYA(from, ts_load->load(from->getParents()));
+      }
+    }
+  }
+
+  outcome::result<std::vector<SignedMessage>> MessagePool::select(
+      TipsetCPtr ts, double ticket_quality) const {
+    OUTCOME_TRY(base_fee, ts->nextBaseFee(env_context.ipld));
+    vm::runtime::Pricelist pricelist{ts->epoch()};
+    OUTCOME_TRY(cached, env_context.interpreter_cache->get(ts->key));
+    vm::state::StateTreeImpl state_tree{env_context.ipld, cached.state_root};
+    auto pending{by_from};
+    constexpr auto kDepth{20};
+    OUTCOME_TRY(path, findPath(env_context.ts_load, head, ts, kDepth));
+    for (auto &ts : path.first) {
+      OUTCOME_TRY(ts->visitMessages(
+          {ipld, false, true},
+          [&](auto, auto bls, auto &cid, auto *smsg, auto *msg)
+              -> outcome::result<void> {
+            if (bls) {
+              auto sig{bls_cache.find(cid)};
+              if (sig != bls_cache.end()) {
+                mpool::add(pending, {*msg, sig->second});
+              }
+            } else {
+              mpool::add(pending, *smsg);
+            }
+            return outcome::success();
+          }));
+    }
+    for (auto &ts : path.second) {
+      OUTCOME_TRY(ts->visitMessages(
+          {ipld, false, true},
+          [&](auto, auto, auto &, auto *, auto *msg) -> outcome::result<void> {
+            mpool::remove(pending, msg->from, msg->nonce);
+            return outcome::success();
+          }));
+    }
+    std::vector<SignedMessage> messages;
+    std::vector<MsgChain::Ptr> chains;
+    GasAmount gas_limit{kBlockGasLimit};
+    auto createChains{
+        [&](auto &chains, auto &from, auto &by_nonce) -> outcome::result<void> {
+          OUTCOME_TRY(actor, state_tree.get(from));
+          append(
+              chains,
+              createMessageChains(
+                  by_nonce, base_fee, actor.nonce, actor.balance, pricelist));
+          return outcome::success();
+        }};
+    // TODO(turuslan): priority addrs
```

Comment:
Do we still need this comment?

Reply:
Yes, these addresses will come from somewhere outside

Reply:
Ok